### ISSUE TITLE
OpenML100 too easy?

### ISSUE THREAD

**Issue by amueller** [openml/OpenML#491]
I'm not sure where would be a good point to discuss this, but I think some of the OpenML100 are too easy.
Why? I looked at the first *ten* run on each of the tasks and checked the best accuracy and AUC, and many of them are around 99% or even 1. There are tens of thousands of runs, but downloading them takes a long time, so I didn't do this here. Feel free to run again with greater size blow. If most of the runs have accuracy around 1, I think the dataset/task is too easy.

```python
import numpy as np
import openml
openml100_tasks = openml.tasks.list_tasks(tag="openml100")

for task in openml100_tasks:
    runs_list = openml.runs.list_runs(task=[task], size=10)
    runs = openml.runs.get_runs(runs_list.keys())
    best_acc = np.max([run.evaluations['predictive_accuracy'] for run in runs])
    best_auc = np.max([run.evaluations['area_under_roc_curve'] for run in runs])

    print(task, best_acc, best_auc)
````

![image](https://user-images.githubusercontent.com/449558/31540230-5c76cc7a-b00b-11e7-938e-9d4d9efacfec.png)

Keep in mind, this is the best of ONLY THE FIRST 10 RUNS!

**Comment by amueller** [comment]
45% of dataset in OpenML100 have a run within auc of 0.99 within the first 10 runs.

**Comment by amueller** [comment]
Better with evaluations:
```python
import numpy as np
import openml
openml100_tasks = openml.tasks.list_tasks(tag="openml100")
res = []

for task in openml100_tasks:
    evaluations = openml.evaluations.list_evaluations("area_under_roc_curve", task=[task], size=100)
    best_auc = np.max([ev.value for ev in evaluations.values()])
    print(task, best_auc)
    res.append([task, best_auc])
```

**Comment by amueller** [comment]
using 10000 runs:
![image](https://user-images.githubusercontent.com/449558/31541337-9dd48416-b00e-11e7-9e8d-ca6f9e7297f9.png)

55% of the datasets have best auc above 0.99

**Comment by ledell** [comment]
@amueller Yeah, I'd say this is concerning and unrealistic.  This also makes the collection unsuitable for benchmarking stacking algorithms.

**Comment by amueller** [comment]
@ledell do you think "best algorithm" is a good criterion? I was also looking at "90th percentile" though I'm not sure if that's any better. Median wasn't as bad.

**Comment by amueller** [comment]
Maybe the AUC is also a problem? On this dataset: https://www.openml.org/t/9955 the best AUC is 0.99, the best accuracy is 0.72. And the distributions are very different. I'm not sure what's happening here.

**Comment by joaquinvanschoren** [comment]
I created this earlier:
https://github.com/openml/Study-14/blob/master/OpenML100%20-%20score%20overview.ipynb

It uses the median rather than the maximum, as I felt that the maximum would return only overfitted results. They also include longer pipelines and stacked classifiers.

Even so, it is clear that some are too easy, especially tamilnadu-electricity, mushroom, collins, irish, and cardiotocography (musk is already removed since it is not iid).

If we remove datasets from the benchmark, there should be a consistent reason to do so, something like 'decision tree performance should be < 0.95 AUC). We want to have a clear reason why some datasets are included and others are not. Currently, I'm not sure what a good measure of being 'too easy' would be. Thoughts?

I'll check the AUC question. Looking at the class distribution of that dataset (https://www.openml.org/d/1492) it a perfectly balanced 100-class problem, with 16 examples per class. From memory, I believe that the evaluation engine uses a weighted 1-vs-all approach. Hence, you evaluate 16 examples from one class vs. 1584 of the other class. No matter what your classifier does, the 'default' AUC (0.5) will be much higher than the default accuracy (0.01). Would that be an explanation?

**Comment by amueller** [comment]
yeah, it could be, but it kind of defies my intuition. Maybe it's just not that intuitive? But that would also be good to know. I.e. that on multi-class with many classes AUC is way higher than accuracy. That should inform which metric to use for selection...

**Comment by joaquinvanschoren** [comment]
After talking with Bernd and Giuseppe, we thought it would be interesting to see the score distribution per dataset. The logic is that datasets with a sharply peaked distribution are not interesting (all algorithms perform about the same), but those that have long tails are very interesting because a few algorithms manage to do better than most.

Results here:
https://github.com/openml/Study-14/blob/master/OpenML100%20-%20score%20overview.ipynb

I made 2 versions: the first uses only the results from the random bot (all R algorithms + random search), so that you have about the same range of algorithms per dataset. However, only about 50% of the OpenML100 datasets are covered by the random bot right now. (I did notice that about 100,000 runs were not evaluated on the server yet, I'm running these overnight and update tomorrow when the server has caught up). The second just uses all available runs (or at least a subset of 2000 runs for now).

You'll see that a few datasets have a single sharp peak. Let's discuss whether these should be removed. 

Many have a long left tail (under/overfitting models?), others are bimodal or have a noticeable right tail.

**Comment by joaquinvanschoren** [comment]
Notebook updated with better plots. This is based on only the random bot runs:
![auc-randombot](https://user-images.githubusercontent.com/1724503/32004957-56bcacd6-b9a3-11e7-8776-80ce655d9441.png)

**Comment by amueller** [comment]
Does the plot smooth over 1.0? It's hard to see if the peak for banknote-authentication is at 1.0 or below that. I also have difficulty reconciling these plots with my numbers...

I totally agree that the long-tail datasets are particularly interesting ones. That's why I looked into https://www.openml.org/t/9955 

But for that one, the results seem to be quite different if you look at accuracy vs AUC.

**Comment by amueller** [comment]
Copy-pasting from the slack channel for the record:
> ok let's say I am concerned about three datasets: https://www.openml.org/d/1466 https://www.openml.org/t/9979 https://www.openml.org/t/34538

**Comment by PhilippPro** [comment]
I think I can possibly explain some things. There are algorithm configurations that are completely useless, these are the results at 0.5. Then we also have several peaks sometimes and that probably is when some algorithms can just not be better than a certain threshold, so e.g. they are near 0.8 while the other algorithm can achieve near 0.95 (with different settings). 

steel-plates-fault seems to be too easy, I think.

**Comment by mfeurer** [comment]
While looking through the datasets once again I found that some of them have a single feature which explains it all. To find all of the datasets in question, I made this little notebook: https://github.com/openml/Study-14/blob/master/openml100_single_feature.ipynb

It can explain some of the incredible performances... However, it does not explain cardiotography and mice_protean (datasets mentioned by @amueller), but at least the latter one has some other issues (see google doc). We might want to exclude all datasets where the RF default configuration of sklearn or mlr gets 100% accuracy. Or spend some more computation time and actually try to find if there are combinations of two features which can perfectly explain the dataset and exclude those.

**Comment by amueller** [comment]
@mfeurer yeah that's similar to the approach I took when I initially posted this. Btw, What I did was train a depth 1 decision tree for each dataset, which is a bit easier ;)

**Comment by amueller** [comment]
(oh and I did that without doing cross-validation, asking the question whether there is a possible solution using one feature - though you could also do it using cross-validation)

**Comment by mfeurer** [comment]
Oh, sorry for redoing the work you already did. Although, if you had more than two classes a decision stump wouldn't help, right?

**Comment by amueller** [comment]
Let me find the issue with the snippet. I used ``max_leafs=n_classes`` or something like that.

**Comment by amueller** [comment]
See #474.

**Comment by mfeurer** [comment]
I see, and the other ones probably aren't there because they're too big?

**Comment by amueller** [comment]
@mfeurer in that other listing? Yeah probably. I ran that on all "small" datasets or openml, not on openml100.

**Comment by joaquinvanschoren** [comment]
New notebook, basically the same as @mfeurer but using decisionstumps instead of randomforests and using training set error instead of a hold out:
https://github.com/openml/Study-14/blob/master/OpenML100%20too%20easy.ipynb

Only 1 dataset (collins) has perfect accuracy.

**Comment by joaquinvanschoren** [comment]
@amueller I did this too see whether we can find formal grounds for removing these simple datasets. Decision stump performance would be one approach. The one-feature approach of @mfeurer also seems reasonable (and seems to catch more). 

Other datasets, like cardiotocography, cannot be removed based on this. They do get perfect scores using simple decision trees of 1NN, but both those algorithms easily overfit, so you would have to use cross-validation scores?

**Comment by mfeurer** [comment]
The issue why my RF-based approach caught more datasets is probably because scikit-learn does not implement proper splitting on categorical features. If the number of leafs is equal to the number of classes, this might not perfectly work. Given that the RF can do as many splits as necessary (or as there is data for) it is more powerful.

**Comment by amueller** [comment]
@mfeurer I don't follow. I would rather expect non-monotonous relationships between a continuous variable and the target that can be modeled by the forest but you need multiple splits.

**Comment by mfeurer** [comment]
Please excuse the unclear wording, here is a clear version:

The reason why my RF-based approach caught more datasets is probably because scikit-learn's decision trees does not implement proper splitting on categorical features. If the number of leafs is equal to the number of classes, the decision tree might not be able to find the splits which result in 100% accuracy. Given that the RF can do as many splits as necessary (or as there is data for) it is more powerful as it can 'fake' correct splits on categorical data.

**Comment by amueller** [comment]
Can you give an example? even if you split off one level each time, which is what scikit-learn can do, that should solve it with n_classes many leaves?

**Comment by mfeurer** [comment]
Not exactly what I was pointing out, but have a look at the attribute `tree` of the [cjs dataset](https://www.openml.org/d/23380). Basically, the number of splits, given that the feature is numerical, you need to make is higher than the number of classes.

**Comment by amueller** [comment]
Ah, got it.

**Comment by joaquinvanschoren** [comment]
I updated the notebook with a new approach:
https://github.com/openml/Study-14/blob/master/OpenML100%20too%20easy.ipynb

This is basically the same as what @mfeurer did, only replacing the holdout with a stratified 10fold CV to be sure that the 100% performance isn't a fluke.

cjs, irish, tamilnadu and collins all get 100% on one feature, and should likely be removed.
mushroom gets 98.6%
cardiotocography gets 47.7% - even though it is generally easy (e.g. using 1NN), you need more than 1 feature.

**Comment by amueller** [comment]
there is two different V1 of collins:
https://www.openml.org/d/478
https://www.openml.org/d/40667
one is active, one is not.
and V2 is binarized. Which version did you run on?

**Comment by amueller** [comment]
Also, that seems to contradict the version + name is unique.

**Comment by amueller** [comment]
I think colins just has the target column in it twice in 478. Or is it also the Corpus variable?

**Comment by joaquinvanschoren** [comment]
Thanks for catching that. Must be a stale search engine record. Gone now.
API returns only these: https://www.openml.org/api/v1/data/list/data_name/collins

**Comment by amueller** [comment]
Ok, but still has the target column twice, right?

**Comment by amueller** [comment]
I don't understand what's happening here actually. X has 21 features, but dataset has 24 features. One (text) is ignored, one is the target attribute. But that means one other one is missing...

**Comment by amueller** [comment]
ah, counter is also excluded... it's really non-trivial to get the correspondence between feature names and features. hopefully #363 will fix that.

**Comment by amueller** [comment]
ok yeah collins has the target duplicated as a feature... time for a new version? ;)

**Comment by amueller** [comment]
For "irish" I couldn't get my hands on the original paper, but it seems like the way the data is used to define a classification problem doesn't make sense, since one of the features completely explains the outcome.

**Comment by amueller** [comment]
cjs also doesn't seem to make sense. We're trying to estimate the effect of treatment, but the target variable is how a tree was treated? (which is unique to each tree). It kind of would make more sense to estimate the number of internodes and see if treatment was an important factor? But (similar to what I see in irish) this looks more like a statistical inference problem than a supervised learning problem. Where did the data come from and how was the default target attribute defined?

**Comment by amueller** [comment]
For the Tamilnadu Electricity Board Hourly Readings dataset no explanation seems to really exist, and the id leaks the label... we could ignore the ID, but given that there's no info on the dataset, it seems a bit questionable.

**Comment by amueller** [comment]
In "mushroom" the "odor" features basically leaks the label.

I wonder if having one of these "simple" ones in there is ok. Basically this is trivial with a tree, but not with a linear model, but I guess that's true whenever a categorical leaks the label.

**Comment by amueller** [comment]
here's some input on "sick":
![image](https://user-images.githubusercontent.com/449558/32635299-8ffafc60-c564-11e7-9adf-49a8096393a2.png)
This is the t3 feature...

**Comment by amueller** [comment]
The MiceProtein 4550 you used is deactivated now?

**Comment by amueller** [comment]
Btw, it goes down from 97% to 95% on the old mouse if you mean over folds instead of max over folds.

**Comment by amueller** [comment]
I would still like to investigate the AUC statistics I started the thread with, but for now I would say remove the double target in collins and it's fine, but maybe remove irish and cjs and Tamilnadu? Do you want me to search for replacements? Theoretically we could ignore the leaking attributes, but for irish and cjs I feel like the targets are really weirdly defined.

**Comment by joaquinvanschoren** [comment]
Collins: Indeed many things wrong here. It only contained part of the data (corpus 1), and the Genre and Corpus together make up the target, so should be ignored. The counter probably as well (it just counts documents within each corpus).

New dataset and task: https://www.openml.org/d/40971

**Comment by joaquinvanschoren** [comment]
MiceProtein: A new version is created, correcting for the 3 last features which leaked target information. New version: https://www.openml.org/d/40966

**Comment by joaquinvanschoren** [comment]
Irish: The education_level kind of gives the target away. Either leave as a really simple dataset (but exclude for benchmarking), or update to ignore that feature. I prefer the former.

cjs: From the description, the natural thing to predict would be the length of the terminal sprouts (TL), the lengths of the individual internodes, and the number of internodes. The way that the data is formatted, though, the individual lenghts are in individual features, filled with missing values if there are fewer than 29 internodes. So TL is the only reasonable target. I could create a new version, but still think that this is too weird for the benchmarks?

Tamilnadu: Definitely ignore the Service ID. Predicting the type of building based on Wattage and Volt-Ampere still seems like an interesting 2D problem. New version?

Mushroom: well know but maybe too simple for benchmarking?

**Comment by joaquinvanschoren** [comment]
Tamilnadu: I'm guessing these are hourly readings per serviceID, so this is actually a time series? Hence, a counter should be added counting the rows for each serviceID?

**Comment by amueller** [comment]
Sorry for slow reply, I think I agree with your assessment. I think cjs is a bit weird indeed.
Ostensibly Tamilnadu is time series, but the fact that there is no explanation whatsoever about the data available is a bit unfortunate.

How were the target attributes for irish and cjs selected when they were first uploaded?

**Comment by joaquinvanschoren** [comment]
The new CC18 benchmark solves these issues.

**Comment by amueller** [comment]
does it? ;)

**Comment by joaquinvanschoren** [comment]
We test whether:
- A decision tree or logistic regression model can get perfect CV performance
- A single feature can perfectly predict the target label (tested by building a random forest on every single feature alone).

If any of these hold, we deem the dataset too easy.

**Comment by amueller** [comment]
Oh cool! I'll check it out and break it later ;)

**Comment by joaquinvanschoren** [comment]
Can't tell if you're being sarcastic or not :)

**Comment by amueller** [comment]
I think these are some of the measures that we discussed and I think they are great additions!

But probably they are not perfect and so I'll see what I find wrong with the current CC-18 and report back what can be improved for CC-20(?) to be even better!

**Comment by PhilippPro** [comment]
Is there already a paper for the CC18 with results?
